akvo/rsr/management/commands/data_integrity.py

```diff
@@ -0,0 +1,181 @@
+        'Parent result project title'
+    ]
+    for result in Result.objects.all().select_related(
+            'project', 'parent_result', 'parent_result__project'
+    ).order_by('project__pk'):
+        if result.parent_result:
+            if result.parent_result.project not in result.project.parents_all():
+                problem_results.append([
+                    result.project.pk,
+                    result.pk,
+                    ", ".join([str(parent.pk) for parent in result.project.parents_all()]),
+                    result.parent_result.project.pk,
+                    result.project.title,
+                    result.title,
+                    result.parent_result.project.title,
+                ])
+
+    if len(problem_results):
+        print "Results where Result.parent_result.project is not in result.project.parents_all()"
+        print problem_results.export('tsv')
+    else:
+        print "No problems with results and their projects"
+    print "\n\n"
+
+
+def inconsistent_indicators():
+    problem_indicators = tablib.Dataset()
+    problem_indicators.headers = [
+        'Project ID',
+        'Indicator ID',
+        'Parent indicator result ID',
+        'Indicator result parent result ID',
+        'Indicator title',
+        'Project title',
+        'Result ID',
+        'Result title',
+    ]
+    for indicator in Indicator.objects.all().select_related(
+            'result__parent_result', 'parent_indicator', 'parent_indicator__result',
+            'result__project',
+    ).order_by('result__project__pk', 'result__pk'):
+        if indicator.parent_indicator:
+            if indicator.parent_indicator.result != indicator.result.parent_result:
+                problem_indicators.append([
+                    indicator.pk,
+                    indicator.parent_indicator.result,
+                    indicator.result.parent_result,
+                    indicator.title,
+                    indicator.result.project.pk,
+                    indicator.result.project.title,
+                    indicator.result.pk,
+                    indicator.result.title,
+                ])
+
+    if len(problem_indicators):
+        print "Indicators where the parent results don't match"
+        print problem_indicators.export('tsv')
+    else:
+        print "No problems with indicators and their results"
+    print "\n\n"
+
+
+def inconsistent_periods():
+    problem_periods = tablib.Dataset()
+    problem_periods.headers = [
+        'Period ID',
+        'Parent period indicator ID',
+        'Indicator parent indicator ID',
+        'Period start and end',
+        'Project ID',
+        'Project title',
+        'Result ID',
+        'Result title',
+        'Indicator ID',
+        'Indicator title',
+    ]
+    print "Indicator periods where parent indicators don't match"
+    print problem_periods.export('csv'),
+    # Note: this double loop is to avoid running out of memory (I think). Previously the loop was
```

Comment:
I think the double loop can be gotten rid of, by moving the if condition checks into the DB. 

```python
for period in IndicatorPeriod.objects.exclude(parent_period=None)\
                                     .exclude(parent_period__indicator__pk=F('indicator__parent_indicator__pk'))
```